crates/red_knot_python_semantic/src/types/call.rs

```diff
@@ -1,423 +1,262 @@
+
+        /// The variants that can be called with the given arguments.
+        callable: Box<[CallBinding<'db>]>,
+
+        /// The union type that we tried calling.
+        called_ty: UnionType<'db>,
     },
+
+    /// The type has a `__call__` method but it isn't always bound.
     PossiblyUnboundDunderCall {
-        callable_ty: Type<'db>,
-        return_ty: Type<'db>,
+        called_type: Type<'db>,
+        outcome: Box<CallOutcome<'db>>,
     },
+
+    /// The type is callable but not with the given arguments.
+    BindingError { binding: CallBinding<'db> },
 }
 
-impl<'db> NotCallableError<'db> {
-    /// The return type that should be used when a call is not callable.
-    pub(super) fn return_type(&self) -> Type<'db> {
+impl<'db> CallError<'db> {
+    /// Returns a fallback return type to use that best approximates the return type of the call.
+    ///
+    /// Returns `None` if the type isn't callable.
+    pub(super) fn return_type(&self, db: &'db dyn Db) -> Option<Type<'db>> {
         match self {
-            Self::Type { return_ty, .. } => *return_ty,
-            Self::UnionElement { return_ty, .. } => *return_ty,
-            Self::UnionElements { return_ty, .. } => *return_ty,
-            Self::PossiblyUnboundDunderCall { return_ty, .. } => *return_ty,
+            CallError::NotCallable { .. } => None,
+            CallError::NotCallableVariants {
+                not_callable,
+                callable,
+                ..
+            } => {
+                // If all variants are not callable, return `None`
+                if callable.is_empty()
+                    && not_callable.iter().all(|binding| binding.binding.is_none())
+                {
+                    None
+                } else {
+                    // If some variants are callable, and some are not, return the union of the return types of the callable variants
+                    // combined with `Type::Unknown`
+                    Some(UnionType::from_elements(
+                        db,
+                        callable.iter().map(bind::CallBinding::return_type).chain(
+                            not_callable
+                                .iter()
+                                .map(NotCallableVariant::unwrap_return_type),
+                        ),
+                    ))
+                }
+            }
+            Self::PossiblyUnboundDunderCall { outcome, .. } => Some(outcome.return_type(db)),
+            Self::BindingError { binding } => Some(binding.return_type()),
         }
     }
 
+    /// Returns the return type of the call or a fallback that
+    /// represents the best guess of the return type (e.g. the actual return type even if the
+    /// dunder is possibly unbound).
+    ///
+    /// If the type is not callable, returns `Type::Unknown`.
+    pub(super) fn unwrap_return_type(&self, db: &'db dyn Db) -> Type<'db> {
+        self.return_type(db).unwrap_or(Type::unknown())
+    }
```

Comment:
I think I'd expect a function with `unwrap` in its name to panic if it contained an error. But this returns a fallback in the case of an error. And in fact, we know we have an error here, since this is a method on `CallError` 😄 Maybe something like

```suggestion
    pub(super) fn fallback_return_type(&self, db: &'db dyn Db) -> Type<'db> {
        self.return_type(db).unwrap_or(Type::unknown())
    }
```

?